### PR TITLE
MultiBinding! (Issue 4565)

### LINKED ISSUE #4565: [Enhancement] MultiBinding
## Summary
Add a `MultiBinding` binding type equivalent to WPF.

## API Changes

```
// new
public class MultiBinding : BindingBase
{
     public Collection<BindingBase> Bindings {get; set;}
     public IMultiValueConverter Converter {get; set;}
}

// new
public interface IMultiValueConverter 
{
     object Convert(object[] values, Type targetType, object parameter);
     object[] ConvertBack(object value, Type[] targetTypes, object parameter);
}
```

## Intended Use Case

Example XAML usage:

```
<Button>
     <Button.IsEnabled>
          <MultiBinding Converter="{StaticResource AllMustBeTrueMultiConverter}">
               <Binding Path="ActionIsAllowed"/>
               <Binding Path="SomeOtherRequirement"/>
          </MultiBinding>
     </Button.IsEnabled>
</Button>
```

Comment - It's challenging to build highly responsive UIs in XAML without `MultiBinding`. Currently there are two workarounds, which even together do not address all cases:

(1) Define a single property in a view model that resolves to a value based on other properties, and bind in XAML to the single property.

First, this makes view models more complex and more difficult to maintain. Suppose a property `ActionIsEnabled ` is defined as:
```
public bool ActionIsEnabled => this.ActionIsAllowed && this.SomeOtherRequirement;
```
Each of the setters for the independent properties (`ActionIsAllowed` and `SomeOtherRequirement`) need to include property change notifications for `ActionIsEnabled` in addition to themselves. If the logic for `ActionIsEnabled` is changed, however, then those other property setters (and/or those of other properties) have to potentially be updated as well. 

Moreover, this assumes the dependencies are all entirely within the same view model. If the independent properties are in different classes than the dependent property, however, there may be no practical way to send a property change notification for the dependent property when the independent properties in the other classes are changed. This workaround therefore doesn't work in all cases.

(2) Wrap a control in a new layer (`Grid`, etc) for each condition.

I've seen this done a lot in Angular using nested `<div>`s for each condition. Leaving aside the inefficiency, this does work for simple cases, the most common one being a cascading property like `IsEnabled` or `IsVisible`, where all the conditions must evaluate to `true`. Anything more complex, however, and this system doesn't work either - for example, if ANY rather than ALL conditions must be true.

`MultiBinding` IMHO provides the most robust and elegant means of specifying complex dependencies between control properties and data.

### PR DESCRIPTION
**Pull request by legistek**
### Description of Change ###

Implements `MultiBinding`, per issue #4565.

fixes #4565

### API Changes ###
Added:
`public sealed class MultiBinding : BindingBase`
`public interface IMultiValueConverter`

`public static readonly object DoNothing { get; }` in `Binding`
`public static readonly object UnsetValue { get; }` in `BindableProperty`
(the last two are used in `IMultiValueConverter` implementations)

### Platforms Affected ### 
All

### Behavioral/Visual Changes ###
No changes to existing applications.

New applications can now use `MultiBinding` to bind a target property to a series of `BindingBase` expressions (including nested `MultiBinding`s) that evaluate to a single value through an `IMultiValueConverter` instance provided by the application. For example:

```
<Button Text="Issue Driver's License">
     <Button.IsEnabled>
          <MultiBinding Converter="{StaticResource AllMustBeTrueMultiConverter}">
               <Binding Path="IsOver16"/>
               <Binding Path="HasPassedExam"/>
               <Binding Path="HasCriminalRecord" Converter="{StaticResource Inverter}"/>
          </MultiBinding>
     </Button.IsEnabled>
</Button>
```

All types of `Binding`s should be valid child bindings for the `MultiBinding`, including implicit source (`BindingContext`), explicit source `Binding`, relative source `Binding`, and even additional `MultiBinding`s.

Except for allowing nested `MultiBinding`s, all efforts have been made to match the WPF spec and behavior for the same feature.

### Testing Procedure ###
Unit tests in Xamarin.Forms.Core

### PR Checklist ###
- [x] Targets the correct branch
- [ ] Tests are passing (or failures are unrelated)

### CODE REVIEW

File: Xamarin.Forms.Core.UnitTests/MultiBindingTests.cs [hunk 1]
```diff
@@ -0,0 +1,872 @@
...
+		{
+			base.TearDown();
+			Device.PlatformServices = null;
+		}
+
+		[Test]
+		public void TestChildOneWayOverridesMultiTwoWay()
+		{
+			// This tests a weird edge case where the MultiBinding is TwoWay but
+			// one of the child bindings is OneWay. This results a situation where
```
Comment (samhouts) [inline Xamarin.Forms.Core.UnitTests/MultiBindingTests.cs]:
```suggestion
			// one of the child bindings is OneWay. This results in a situation where
```

File: Xamarin.Forms.Core.UnitTests/MultiBindingTests.cs [hunk 2]
```diff
@@ -0,0 +1,872 @@
...
+
+			// Martial law is over
+			group.SuspendAll = false;
+			Assert.IsTrue(checkBox.IsChecked);
+
+			// But she got in an accident and now can't drive again
+			group.Person5.IsSuspended = true;
+			Assert.IsFalse(checkBox.IsChecked);
+
+			// The new PM has pardoned everyone after the end of the rebellion
```
Comment (samhouts) [inline Xamarin.Forms.Core.UnitTests/MultiBindingTests.cs]: so nice when a unit test has a happy ending

File: Xamarin.Forms.Core/MultiBinding.cs [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+			{
+				ThrowIfApplied();
+				_converterParameter = value;
+			}
+		}
+
+		public Collection<BindingBase> Bindings
+		{
+			get => _bindings ?? (_bindings = new Collection<BindingBase>());
+			set => _bindings = value;
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/MultiBinding.cs]: it should probably throw if applied, and also throw if the collection is modified once the binding is applied

File: Xamarin.Forms.Core/MultiBinding.cs [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+		{
+			get { return _converterParameter; }
+			set
+			{
+				ThrowIfApplied();
+				_converterParameter = value;
+			}
+		}
+
+		public Collection<BindingBase> Bindings
```
[the remark this answers is not in the document]
Comment (StephaneDelcroix) [reply Xamarin.Forms.Core/MultiBinding.cs]: or an IList if you're not tracking INCC

File: Xamarin.Forms.Core/MultiBinding.cs [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+				ConverterParameter = ConverterParameter,
+				Bindings = new Collection<BindingBase>(this.Bindings),
+				FallbackValue = FallbackValue,
+				Mode = Mode,
+				TargetNullValue = TargetNullValue,
+				StringFormat = StringFormat,
+			};
+		}
+
+		internal override void Apply(bool fromTarget)
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/MultiBinding.cs]: if I understand correctly, the Converter is mandatory, so it should be nice to Log a message if the converter is missing

**Comment by AmrAlSayed0** [comment]
Is MultiBinding going to be in 4.5.0?

**Comment by legistek** [comment]
@AmrAlSayed0 it has not been abandoned but I have been swamped with my actual job. I'm fine with @StephaneDelcroix's changes so if he or anyone wants to just make them and run with this that's fine by me. Otherwise I'll make them when I'm able.

File: Xamarin.Forms.Core/MultiBinding.cs (hunk 3 above)
Comment (legistek) [reply Xamarin.Forms.Core/MultiBinding.cs]: Ok on throw if applied. To detect collection modified wouldn't it have to be an INCC though? Is it worth it?

Comment (legistek) [reply Xamarin.Forms.Core/MultiBinding.cs]: > I know there's none in WPF, but what would be a good markup extension for this ? (we can bring that on a different PR)

I was thinking about that myself, but I don't know of any precedent for a collection being represented through a markup extension, do you?

**Comment by legistek** [comment]
@StephaneDelcroix I think I'm finished here, just not sure if you actually want me to bother trying to detect binding collection changes through INCC. 

This should now be properly rebased on 4.6.0 too btw.

**Comment by neobenedict** [comment]
Error		Position 56:34. Type MultiBinding not found in xmlns http://xamarin.com/schemas/2014/forms

Version 4.6.0.726+446-sha.d0dc59903-azdo.3681414

Is this not usable yet?

**Comment by samhouts** [comment]
This will be part of 4.7.0. You may try it in the latest nightly build. Thanks!

**Comment by ncarandini** [comment]
I've made a [Demo](https://github.com/ncarandini/XFMultiBindingDemo) and it works.
For comparison and for those who don't like using converters and multi bindings I've added an example on how to create a property in the ViewModel that does the same thing.
